src/workspace.rs

```diff
@@ -0,0 +1,393 @@
+    }
+}
+
+impl TryFrom<&PathBuf> for ManifestPath {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &PathBuf) -> Result<Self, Self::Error> {
+        ManifestPath::new(value)
+    }
+}
+
+impl Default for ManifestPath {
+    fn default() -> ManifestPath {
+        ManifestPath::new(MANIFEST_FILE).expect("it's a valid manifest file")
+    }
+}
+
+impl AsRef<Path> for ManifestPath {
+    fn as_ref(&self) -> &Path {
+        self.path.as_ref()
+    }
+}
+
+/// Create, amend and save a copy of the specified `Cargo.toml`.
+pub struct Manifest {
+    path: ManifestPath,
+    toml: value::Table,
+}
+
+impl Manifest {
+    /// Create new CargoToml for the given manifest path.
+    ///
+    /// The path *must* be to a `Cargo.toml`.
+    pub fn new<P>(path: P) -> Result<Manifest>
+    where
+        P: TryInto<ManifestPath, Error = anyhow::Error>,
+    {
+        let manifest_path = path.try_into()?;
+        let toml = fs::read_to_string(&manifest_path).context("Loading Cargo.toml")?;
+        let toml: value::Table = toml::from_str(&toml)?;
+
+        Ok(Manifest {
+            path: manifest_path,
+            toml,
+        })
+    }
+
+    /// Get mutable reference to `[lib] crate-types = []` section
+    fn get_crate_types_mut(&mut self) -> Result<&mut value::Array> {
+        let lib = self
+            .toml
+            .get_mut("lib")
+            .ok_or(anyhow::anyhow!("lib section not found"))?;
+        let crate_types = lib
+            .get_mut("crate-type")
+            .ok_or(anyhow::anyhow!("crate-type section not found"))?;
+
+        crate_types
+            .as_array_mut()
+            .ok_or(anyhow::anyhow!("crate-types should be an Array"))
+    }
+
+    /// Add a value to the `[lib] crate-types = []` section.
+    ///
+    /// If the value already exists, does nothing.
+    pub fn with_added_crate_type(&mut self, crate_type: &str) -> Result<&mut Self> {
+        let crate_types = self.get_crate_types_mut()?;
+        if !crate_type_exists(crate_type, crate_types) {
+            crate_types.push(crate_type.into());
+        }
+        Ok(self)
+    }
+
+    /// Remove a value from the `[lib] crate-types = []` section
+    ///
+    /// If the value does not exist, does nothing.
+    pub fn with_removed_crate_type(&mut self, crate_type: &str) -> Result<&mut Self> {
+        let crate_types = self.get_crate_types_mut()?;
+        if crate_type_exists(crate_type, crate_types) {
+            crate_types.retain(|v| v.as_str().map_or(true, |s| s != crate_type));
+        }
+        Ok(self)
+    }
+
+    /// Replace relative paths with absolute paths with the working directory.
+    ///
+    /// Enables the use of a temporary amended copy of the manifest.
+    ///
+    /// # Rewrites
+    ///
+    /// - `[lib]/path`
+    /// - `[dependencies]`
+    ///
+    /// Dependencies with package names specified in `exclude_deps` will not be rewritten.
+    fn rewrite_relative_paths<I, S>(&mut self, exclude_deps: I) -> Result<&mut Self>
```

Comment:
Wow this was way more work than I expected. O.O

Reply:
Me too...